compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3144,6 +3144,125 @@ static inline void loadArrayCmpSources(TR::Node *node, TR::InstOpCode::Mnemonic
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb, node, vec0Reg, vec0Reg, vec1Reg);
+   // residue end
+
+   // result
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, resultLabel);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vec0Reg, vec0Reg, vec0Reg);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::vclzlsbb, node, temp2Reg, vec0Reg);
+   
+   // vec-eq-mask = -1 (if equal) | 0 (it not-equal)
+   generateShiftLeftImmediateLong(cg, node, tempReg, temp2Reg, 59);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, tempReg, tempReg, 63);
+
+   // offset = matched-byte-count & ~(vec-eq-mask) + remainder & (vec-eq-mask)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, temp2Reg, temp2Reg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, returnReg, returnReg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, temp2Reg);
+
+   // index = index + offset, if we need to return unmatched index, then we are done here 
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, indexReg, returnReg);
+   
+   if (!node->isArrayCmpLen())
+   {
+      // if it was a match then we need index -= 1 to get last index for compare
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, tempReg);
+
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, src1AddrReg, src1AddrReg, returnReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, src2AddrReg, src2AddrReg, returnReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, src1AddrReg, src2AddrReg);
+      // result = -1,0,1
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::setb, node, temp2Reg, condReg);
+      // convert -1,0,1 to 1,0,2 to match current arraycmp return value
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, temp2Reg, temp2Reg);
+      generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, returnReg, temp2Reg, 2, 3);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, temp2Reg);
+   }
+
+   int32_t numRegs = 9;
+
+   TR::RegisterDependencyConditions *dependencies = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numRegs, cg->trMemory());
+   dependencies->addPostCondition(src1AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(src2AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(tempReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(returnReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(condReg, TR::RealRegister::cr6);
+   dependencies->addPostCondition(vec0Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(vec1Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(indexReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(temp2Reg, TR::RealRegister::NoReg);
```

Comment:
as it is, tempReg, indexReg, src1AddrReg, and src2AddrReg have to be marked as excludeGPR0, since they are used in various places as RA in addi or load (where r0 means 0 instead of value in r0). You can certainly rearrange code to reduce the number of such setting the flag. 

At the same time, you even don't need to modify src1AddrReg and src2AddrReg at all, such that you can share the registers from the respective node and eliminate the potential register-copy as result of gprClobberEvaluate.

Reply:
In previous WIP implementation, I didn't modify the address registers. I will change it to use other available registers.

Reply:
Removed `gprClobberEvaluate` from src1Addr, src2Addr. Added excludeGPR0 to tempReg and indexReg.